source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,296 @@
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                message_.memory_semantics_operand_index() == 0 ? 2 : 3))
+            ->GetSingleWordInOperand(0);
+  }
+  auto first_5bits_new_memory_semantics =
+      static_cast<SpvMemorySemanticsMask>(new_memory_sematics_value & 0x1F);
+  auto first_5bits_old_memory_semantics =
+      static_cast<SpvMemorySemanticsMask>(old_memory_sematics_value & 0x1F);
+  auto memory_model = static_cast<SpvMemoryModel>(
+      ir_context->module()->GetMemoryModel()->GetSingleWordInOperand(1));
+
+  if (!IsValidConverstion(needed_atomic_instruction->opcode(),
+                          first_5bits_old_memory_semantics,
+                          first_5bits_new_memory_semantics, memory_model)) {
+    return false;
+  }
+
+  // The higher bits value of old and new memory semantics id must be equal.
+  auto higher_bits_new_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      new_memory_sematics_value & 0xFFFFFFE0);
+  auto higher_bits_old_memory_semantics = static_cast<SpvMemorySemanticsMask>(
+      old_memory_sematics_value & 0xFFFFFFE0);
+  if (higher_bits_new_memory_semantics != higher_bits_old_memory_semantics) {
+    return false;
+  }
+
+  // Instructions that takes two memory semantics, one of the value must be
+  // stronger than the other.
+  if (needed_atomic_instruction->opcode() == SpvOpAtomicCompareExchange ||
+      needed_atomic_instruction->opcode() == SpvOpAtomicCompareExchangeWeak) {
+    auto other_memory_semantics_index = static_cast<SpvMemorySemanticsMask>(
+        ir_context->get_def_use_mgr()
+            ->GetDef(needed_atomic_instruction->GetSingleWordInOperand(
+                message_.memory_semantics_operand_index() == 1 ? 2 : 3))
+            ->GetSingleWordInOperand(0) &
+        0x1F);
+
+    if (first_5bits_new_memory_semantics == other_memory_semantics_index) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void TransformationChangingMemorySemantics::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto needed_atomic_instruction =
+      FindInstruction(message_.atomic_instruction(), ir_context);
+
+  uint32_t needed_index = 0;
+  if (needed_atomic_instruction->opcode() == SpvOpMemoryBarrier) {
+    needed_index = 1;
+  } else {
+    needed_index = message_.memory_semantics_operand_index() == 0 ? 2 : 3;
+  }
+
+  uint32_t higher_bits_memory_semantics =
+      needed_atomic_instruction->GetSingleWordInOperand(needed_index) &
+      0xFFFFFFE0;
+
+  auto new_memory_semantics_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_semantics_new_value_id());
+  // New memory semantics value is OR-ed with the second 27bits of old memory
+  // semantics value.
+  uint32_t total_value =
+      new_memory_semantics_instruction->GetSingleWordInOperand(0) |
+      higher_bits_memory_semantics;
+  new_memory_semantics_instruction->SetInOperand(0, {total_value});
```

Comment:
I don't think this looks correct. If so, your tests should not pass. So rather than explain, I will just move on to your tests.

Reply:
Yes, you are right.
Removed.